segmentstore/server/src/main/java/io/pravega/segmentstore/server/containers/ContainerEventProcessorImpl.java

```diff
@@ -0,0 +1,500 @@
+            Function<List<BufferView>, CompletableFuture<Void>> noHandler = l -> CompletableFuture.completedFuture(null);
+            return segmentSupplier.apply(name)
+                    .thenApply(segment -> {
+                        Exceptions.checkNotClosed(this.closed.get(), this);
+                        EventProcessorImpl processor = new EventProcessorImpl(name, segment, noHandler, config, onClose, executor);
+                        eventProcessorMap.put(name, processor);
+                        return processor;
+                    });
+        }
+    }
+
+    class EventProcessorImpl extends ContainerEventProcessor.EventProcessor {
+
+        private final DirectSegmentAccess segment;
+        private final AtomicLong lastIterationLatency;
+        private final SegmentStoreMetrics.EventProcessor metrics;
+        private final AtomicBoolean closed;
+        private final AtomicBoolean failedIteration;
+
+        private final ProcessorEventData.ProcessorEventDataSerializer serializer = new ProcessorEventData.ProcessorEventDataSerializer();
+
+        //region Constructor
+
+        public EventProcessorImpl(@NonNull String name, @NonNull DirectSegmentAccess segment,
+                                  @NonNull Function<List<BufferView>, CompletableFuture<Void>> handler,
+                                  @NonNull EventProcessorConfig config, @NonNull Runnable onClose,
+                                  @NonNull ScheduledExecutorService executor) {
+            super(String.format("EventProcessor[%d-%s]", containerId, name), executor, name, handler, config, onClose);
+            this.segment = segment;
+            this.lastIterationLatency = new AtomicLong(0);
+            this.metrics = new SegmentStoreMetrics.EventProcessor(name, containerId);
+            this.closed = new AtomicBoolean(false);
+            this.failedIteration = new AtomicBoolean(false);
+        }
+
+        //endregion
+
+        //region EventProcessor implementation
+
+        @Override
+        @SneakyThrows(IOException.class)
+        public CompletableFuture<Long> add(@NonNull BufferView event, Duration timeout) throws TooManyOutstandingBytesException {
+            Preconditions.checkArgument(event.getLength() > 0);
+            Preconditions.checkArgument(event.getLength() + Integer.BYTES < ProcessorEventData.MAX_EVENT_SIZE);
+            Exceptions.checkNotClosed(this.closed.get(), this);
+            // If the EventProcessor reached the limit of outstanding bytes, throw accordingly.
+            if (getOutstandingBytes() > getConfig().getMaxProcessorOutstandingBytes()) {
+                throw new TooManyOutstandingBytesException(this.traceObjectId);
+            }
+
+            ProcessorEventData processorEvent = ProcessorEventData.builder().data(event).build();
+            return this.segment.append(this.serializer.serialize(processorEvent), null, timeout)
+                          .thenApply(offset -> getOutstandingBytes());
```

Comment:
`getOutstandingBytes` invokes `segment.getInfo`; that is relatively expensive as it needs to look up the segment's metadata and it may need to acquire some locks. 

Is there anyway we can avoid this? Furthermore, `getOutstandingBytes` invokes `getInfo` twice in it.

Reply:
I have removed several calls to `segment.getInfo()`:
- `getOutstandingBytes()` calls `segment.getInfo()` only once.
- Added a new field in `EventProcessorImpl` called `segmentStartOffset` that will track the start offset of the Segment. This prevents a couple of calls to `segment.getInfo().getStartOffset()` we did for reads and truncations.

However, I need to use this call sometimes, as I don't know the actual amount of bytes written to the Segment since the introduction of `VersionedSerializer`.